1. Unit testing/1. Basics/Task 01 - Course registration/ClassRegistrationService.test.ts

```diff
@@ -0,0 +1,280 @@
+        id: "C4",
+        code: "CS301",
+        name: "Coures with 42069 credit hours",
+        creditHours: 42069,
+        availableSeats: 1,
+        prerequisites: [],
+        schedule: [
+          {
+            day: "Friday",
+            startTime: "11:00",
+            endTime: "12:30",
+            location: "Room 201",
+          },
+        ],
+      },
+    ];
+    const students: Student[] = [
+      {
+        id: "1",
+        name: "Alice Johnson",
+        completedCourses: ["C2"],
+        currentCourses: [],
+        maxCreditHours: 18,
+      },
+      {
+        id: "2",
+        name: "Bob Smith",
+        completedCourses: [],
+        currentCourses: ["C1"],
+        maxCreditHours: 15,
+      },
+      {
+        id: "3",
+        name: "Charlie Brown",
+        completedCourses: ["C1", "C2"],
+        currentCourses: ["C3"],
+        maxCreditHours: 12,
+      },
+    ];
+
+    courseRegistrationService = new CourseRegistrationService(
+      courses,
+      students
+    );
+  });
+
+  it.each([
+    {
+      scenario: "student not found",
+      studentId: "420",
+      courseId: "C2",
+      success: false,
+      regex: /not found/i,
+    },
+    {
+      scenario: "course not found",
+      studentId: "1",
+      courseId: "C69",
+      success: false,
+      regex: /not found/i,
+    },
+    {
+      scenario: "student already registered",
+      studentId: "3",
+      courseId: "C3",
+      success: false,
+      regex: /registered/i,
+    },
+    {
+      scenario: "student not met prerequisites",
+      studentId: "2",
+      courseId: "C3",
+      success: false,
+      regex: /missing prerequisites/i,
+    },
+    {
+      scenario: "schedule conflicts",
+      studentId: "2",
+      courseId: "C2",
+      success: false,
+      regex: /conflict/i,
+    },
+    {
+      scenario: "course has no available seats",
+      studentId: "1",
+      courseId: "C3",
+      success: false,
+      regex: /available seats/i,
+    },
+    {
+      scenario: "exceeding credit hours limit",
+      studentId: "1",
+      courseId: "C4",
+      success: false,
+      regex: /exceed/i,
+    },
+  ])(
+    "should fail because $scenario",
+    ({ studentId, courseId, success, regex }) => {
+      const result = courseRegistrationService.registerForCourse(
+        studentId,
+        courseId
+      );
+
+      expect(result.success).toBe(success);
+      expect(result.message).toMatch(regex);
+    }
+  );
+
+  it("should register student successfully", () => {
+    const studentId = "1";
+    const courseId = "C1";
+    const student = courseRegistrationService.getStudent(studentId)!;
+    const course = courseRegistrationService.getCourse(courseId)!;
+    const previousCourseAvailableSeats = course?.availableSeats!;
+
+    const result = courseRegistrationService.registerForCourse(
+      studentId,
+      courseId
+    );
+
+    expect(result.success).toBe(true);
+    expect(result.message).toMatch(/successfully/i);
```

Comment:
```suggestion
    expect(result.message).toMatch(/success/i);
```
Nitpick, może tutaj `/success/i` zamiast `/successfully/i` bo z tym np. "Success! You're registered for CS101" nie przechodzi, mimo że to też niby mogła by być poprawna wiadomość zwrotna. Z `/success/i` przechodzi zarówno aktualna wiadomość jak i ta alternatywna

Reply:
fair point